File: dot_product.R
# Dot product in R using a loop and a vector summation
# USAGE:  Rscript dot_product.R


   # Allocate space for and initialize the array

n <- 100000
x <- vector( "double", n )
y <- vector( "double", n )

for( i in 1:n )
{
   x[i] <- as.double(i)
   y[i] <- as.double(3*i)
}


   # Time a simple dot product loop first

dummy <- matrix( 1:125000000 )       # Clear the cache buffers before timing

t_start <- proc.time()[[3]]

dot_product <- 0.0
for( i in 1:n )
{
   dot_product <- dot_product + x[i] * y[i]
}

t_end <- proc.time()[[3]]

print(sprintf("Dot product by for loop took %6.3f seconds", t_end-t_start))
print(sprintf("dot_product = %.6e for vector size %i", dot_product, n ) )


   # Time the dot product by vector sums

dummy <- matrix( 1:125000000 )       # Clear the cache buffers before timing

t_start <- proc.time()[[3]]

dot_product <- sum( x * y )

t_end <- proc.time()[[3]]

print(sprintf("Dot product by vector sum took %6.3f seconds", t_end-t_start))
print(sprintf("dot_product = %.6e for vector size %i", dot_product, n ) )

File: dot_product.m
% Dot product in Matlab using serial for, parfor, and builtin methods
% The builtin parallel routine will use all cores available


   % Allocate and initialize the X and Y vectors

%N=100000;
N=100000000;
ncores=8;

X=rand( N, 1 );
Y=rand( N, 1 );

zarray = zeros( [125000000,1], 'double' );  % Clear the cache

   % Time a serial dot product

Dprod=0.0;
tic  % start the timer

for i = 1:N
   Dprod = Dprod + X(i) * Y(i);
end

t_end = toc;
disp(['serial loop took ', num2str(t_end), ' seconds with Dprod = ', num2str(Dprod)]);


   % Time the internal dot prodduct which is multi-threaded

zarray = zeros( [125000000,1], 'double' );  % Clear the cache
tic  % start the timer

Dprod = dot( X, Y);

t_end = toc;
disp(['builtin dot product took ', num2str(t_end), ' seconds with Dprod = ', num2str(Dprod)]);


   % Time a single-node multi-process  parfor dot product

tic

parpool('local', ncores)    % Fire up the virtual cluster of processes

t_end = toc;
disp([num2str(t_end), ' seconds to spin up multi-process pool']);

zarray = zeros( [125000000,1], 'double' );  % Clear the cache
tic  % start the timer

Dprod=0.0;
parfor i = 1:N
   Dprod = Dprod + X(i) * Y(i);
end

t_end = toc;
disp(['multi-process parfor loop took ', num2str(t_end), ' seconds with Dprod = ', num2str(Dprod)]);


exit; % NOTE: Multi-threads not supported in older versions of Matlab


   % Time a single-node multi-thread parfor matrix multiply

delete( gcp( 'nocreate' ) );   % Destroy the local parpool

tic

parpool('threads', ncores)

t_end = toc;
disp([num2str(t_end), ' seconds to spin up multi-thread pool']);

zarray = zeros( [125000000,1], 'double' );  % Clear the cache
tic  % start the timer

Dprod = 0.0;
parfor i = 1:N
   Dprod = Dprod + X(i) * Y(i);
end

t_end = toc;
disp(['multi-threaded parfor loop took ', num2str(t_end), ' seconds with Dprod = ', num2str(Dprod)]);

exit;


